include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,328 @@
+
+						auto* inDataAdress = inData + readBlockArrayOffset;
+						const void* inSourcePixels[maxPlanes] = { inDataAdress, nullptr, nullptr, nullptr };
+
+						double decodeBuffer[maxChannels] = {}; // ASCT TODO?
+						double swizzledBuffer[maxChannels] = {}; // ASCT TODO?
+
+						for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+							for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+							{
+								impl::CSwizzleableAndDitherableFilterBase<false, false, Swizzle, IdentityDither>::onDecode(inFormat, state, inSourcePixels, decodeBuffer, swizzledBuffer, blockX, blockY);
+
+								const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
+								float* data = reinterpret_cast<float*>(state->scratchMemory + offset);
+
+								auto& [xDecode, yDecode, zDecode] = std::make_tuple(*swizzledBuffer, *(swizzledBuffer + 1), *(swizzledBuffer + 2));
+
+								*data = -xDecode / zDecode;
+								*(data + 1) = -yDecode / zDecode;
+							}
+					};
+
+					IImage::SSubresourceLayers subresource = { static_cast<IImage::E_ASPECT_FLAGS>(0u), state->inMipLevel, state->inBaseLayer, 1 };
+					CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { state->inOffset,state->extent };
+					CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, inFormat);
+
+					auto& inRegions = state->inImage->getRegions(state->inMipLevel);
+					CBasicImageFilterCommon::executePerRegion(state->inImage, decodeAndDivide, inRegions.begin(), inRegions.end(), clipFunctor);
+				}
+
+				{
+					auto getScratchPixel = [&](core::vector4di32_SIMD readBlockPos) -> CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::decodeType*
+					{
+						const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(readBlockPos.x, readBlockPos.y, readBlockPos.z, 0), scratchByteStrides); // TODO
+						return reinterpret_cast<CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase::decodeType*>(reinterpret_cast<uint8_t*>(state->scratchMemory) + scratchOffset);
+					};
+
+					auto computeMaxAbs = [&](core::vectorSIMDi32 readBlockPos) -> void
+					{
+						auto* current = getScratchPixel(readBlockPos);
+						auto& [x, y] = std::make_tuple(*current, *(current + 1));
+
+						auto absoluteX = core::abs(x);
+						auto absoluteY = core::abs(y);
+
+						if (xMaxDecodeAbsValue < absoluteX)
+							xMaxDecodeAbsValue = absoluteX;
+
+						if (yMaxDecodeAbsValue < absoluteY)
+							yMaxDecodeAbsValue = absoluteY;
+					};
+
+					{
+						core::vector3du32_SIMD localCoord;
+						for (auto& z = localCoord[2] = 0u; z < state->extent.depth; ++z)
+							for (auto& y = localCoord[1] = 0u; y < state->extent.height; ++y)
+								for (auto& x = localCoord[0] = 0u; x < state->extent.width; ++x)
+									computeMaxAbs(core::vectorSIMDu32(x, y, z));
+					}
```

Comment:
dont pass over the memory again
https://github.com/Devsh-Graphics-Programming/Nabla/pull/9/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#r609047307